internal/controller/instaslice_controller.go

```diff
@@ -743,3 +759,78 @@ func (r *InstasliceReconciler) ReconcileSCC(ctx context.Context) error {
 	sccs.Client = mfc.NewClient(r.Client)
 	return sccs.Apply()
 }
+
+func (r *InstasliceReconciler) ensureDaemonSet(ctx context.Context) error {
+	log := logr.FromContext(ctx)
+	daemonSet := &appsv1.DaemonSet{}
+	err := r.Get(ctx, types.NamespacedName{Name: InstasliceDaemonsetName, Namespace: InstaSliceOperatorNamespace}, daemonSet)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			daemonSet = r.createInstaSliceDaemonSet(InstaSliceOperatorNamespace)
+			if err := r.Create(ctx, daemonSet); err != nil {
+				log.Error(err, "Failed to create DaemonSet")
+				return err
+			}
+			log.Info("DaemonSet created successfully")
+		} else {
+			log.Error(err, "Failed to get DaemonSet")
+			return err
+		}
+	}
+
+	pollInterval := 5 * time.Second
+	timeout := 2 * time.Minute
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	log.Info("Waiting for at least one DaemonSet pod to be ready...")
+
+	err = wait.PollUntilContextTimeout(ctxWithTimeout, pollInterval, timeout, true, func(ctx context.Context) (bool, error) {
+		isReady, err := r.isDaemonSetPodReady(ctx, daemonSet)
+		if err != nil {
+			log.Error(err, "Failed to check DaemonSet pod readiness")
+			return false, err
+		}
+		return isReady, nil
+	})
+
+	if err != nil {
+		log.Error(err, "Timeout waiting for DaemonSet pod readiness")
+		return err
+	}
+
+	log.Info("At least one DaemonSet pod is ready")
+	return nil
+}
+
+func (r *InstasliceReconciler) isDaemonSetPodReady(ctx context.Context, daemonSet *appsv1.DaemonSet) (bool, error) {
+	var podList v1.PodList
+	labelSelector := labels.SelectorFromSet(daemonSet.Spec.Selector.MatchLabels)
+
+	listOptions := &client.ListOptions{
+		LabelSelector: labelSelector,
+		Namespace:     daemonSet.Namespace,
+	}
+
+	if err := r.List(ctx, &podList, listOptions); err != nil {
+		return false, err
+	}
+
+	for _, pod := range podList.Items {
+		if pod.Status.Phase == v1.PodRunning && len(pod.Status.ContainerStatuses) > 0 && pod.Status.ContainerStatuses[0].Ready {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// Enable creation of controller caches to talk to the API server in order to perform
+// object discovery in SetupWithManager
+func (r *InstasliceReconciler) setupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}).Named("InstaSlice-controller").
+		Watches(&inferencev1alpha1.Instaslice{}, handler.EnqueueRequestsFromMapFunc(r.podMapFunc)).
```

Comment:
I suppose we need to add `Owns()` as well here as it creates and owns the daemonset. Thoughts?

Reply:
That would be disruptive; we want the daemonset pod to be decoupled from the controller.